internal/testutil/log.go

```diff
@@ -0,0 +1,111 @@
+package testutil
+
+import (
+	"encoding/json"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.uber.org/zap/zaptest"
+)
+
+const (
+	logLevelKey   = "level"
+	logMessageKey = "msg"
+	logTimeKey    = "ts"
+)
+
+// LogEntry represents single [zap.Logger] entry.
+type LogEntry struct {
+	Level   zapcore.Level
+	Message string
+	// Integer values are represented as [json.Number].
+	Fields map[string]any
+}
+
+// LogBuffer is a memory buffer for [zap.Logger] entries.
+type LogBuffer struct {
+	t testing.TB
+	b zaptest.Buffer
+}
+
+// NewBufferedLogger returns buffered logger for testing.
+//
+// Entries with severity less than minLevel are never written.
+func NewBufferedLogger(t testing.TB, minLevel zapcore.Level) (*zap.Logger, *LogBuffer) {
+	// TODO: https://github.com/nspcc-dev/neofs-node/issues/3313
+	var lb LogBuffer
+	lb.t = t
+
+	encCfg := zap.NewProductionEncoderConfig()
+	encCfg.LevelKey = logLevelKey
+	encCfg.MessageKey = logMessageKey
+	encCfg.TimeKey = logTimeKey
+
+	zc := zapcore.NewCore(
+		zapcore.NewJSONEncoder(encCfg),
+		zap.CombineWriteSyncers(&lb.b),
+		minLevel,
+	)
+
+	return zap.New(zc), &lb
+}
+
+// AssertEmpty asserts that log is empty.
+func (x *LogBuffer) AssertEmpty() {
+	x.AssertEqual(nil)
+}
+
+// AssertSingle asserts that log has given entry only.
+func (x *LogBuffer) AssertSingle(e LogEntry) {
+	x.AssertEqual([]LogEntry{e})
+}
+
+// AssertEqual asserts that log has exact entries.
```

Comment:
can you add a few words that the order matters? i guess we may have smth like `Match` in the future for parallel testing logs when the order is not defined

Reply:
sure, done